### DISCUSSION TITLE
Hard dependency on yyjson

### DISCUSSION THREAD

**Discussion by ceamac** [fastfetch-cli/fastfetch#523] [General]
So far fastfetch had no mandatory libs.  All of them were dlopened and used only if found at runtime.  Now yyjson is linked in, and fastfetch does not start if it is absent.  (As a side note, you don't even get a link error, it just creates a broken executable).

I don't have a strong opinion about this, it is not wrong per se to depend on a library, but it changes something that was almost a tradition.  So maybe start a discussion about dropping yyjson as hard requirement?  Is it doable?  Is it worth it?

**Comment by CarterLi** [comment]
yyjson should be static linked into fastfetch binary. If not, I'll fix it.

**Comment by ceamac** [comment]
Oh, it was meant to be statically linked?  Then my point it moot.  I'll look into the build system to see why it is like this.  Thanks!

```
$ ldd /mnt/tmp/portage/app-misc/fastfetch-2.0.0/image/usr/bin/fastfetch
	linux-vdso.so.1 (0x00007fff567a8000)
	libyyjson.so.0.7.0 => /usr/lib64/libyyjson.so.0.7.0 (0x00007f3353960000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f3353884000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f33536ab000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3353a92000)
```

**Comment by CarterLi** [comment]
I just checked the binary built by Github Action, it does not depend on yyjson.so.

The source code of yyjson is automatically fetched by CMake FetchContent.

https://github.com/fastfetch-cli/fastfetch/blob/a85da5401df06c14672923cfc2348bb4ed796705/CMakeLists.txt#L40

Do you have yyjson pre-installed? Try removing it and build fastfetch again

**Comment by ceamac** [comment]
Downloading stuff during build it not allowed in Gentoo, the sandbox won't allow it, so I had to remove that part and preinstall yyjson.

**Reply by CarterLi** [reply]
There are many packages have source code dependency of 3rd party projects (eg. git submodules).  How do you handle them?

**Reply by ceamac** [reply]
The same.  Add the dependency as a separate package, then depend on it.  Sometimes we can download 2 or more packages and move them around in src to match their intended places, but it's cleaner this way.  We do that only when needed, for example when a package patches a dependency.

**Reply by CarterLi** [reply]
I see. FetchContent was used to avoid potential license concerns. However as [it has been explicitly permitted by the yyjson author @ibireme](https://github.com/ibireme/yyjson/issues/113), I think it should be okey to embed yyjson source code into fastfetch now.

CC @LinusDierheimer

**Comment by CarterLi** [comment]
[Done](https://github.com/fastfetch-cli/fastfetch/commit/cd4a5122918aeb3f3dd34c4463d3f6f19a882851)

**Reply by ceamac** [reply]
I compiled from the dev branch, it works:
```
$ fastfetch -v
fastfetch 2.0.0-508 (x86_64)
$ lddtree /usr/bin/fastfetch 
fastfetch => /usr/bin/fastfetch (interpreter => /lib64/ld-linux-x86-64.so.2)
    libm.so.6 => /lib64/libm.so.6
        ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2
    libc.so.6 => /lib64/libc.so.6
```

Thanks!

**Comment by CarterLi** [comment]
Hello @ceamac 

Fastfetch v2.0.1 (stable) has been released, with an [optional cmake option `-DENABLE_SYSTEM_YYJSON`](https://github.com/fastfetch-cli/fastfetch/blob/157be8443d2fb2edae791efd616bcce40aa036ea/CMakeLists.txt#L610) which makes fastfetch find system-provided yyjson package (disabled by default). Please have a check.

**Reply by ceamac** [reply]
Hi, I saw, [I packaged it](https://github.com/gentoo/gentoo/blob/master/app-misc/fastfetch/fastfetch-2.0.1.ebuild).

I chose to use external yyjson for now, since it is possible, and removed the patch.

Thank you!

**Reply by CarterLi** [reply]
Please note that yyjson is still hard-linked into fastfetch if `-DENABLE_SYSTEM_YYJSON` is enabled
